token-swap/program/src/curve/stable.rs

```diff
@@ -412,20 +434,84 @@ mod tests {
-        let source_amount: u128 = 100;
-        let swap_source_amount: u128 = 1000;
-        let swap_destination_amount: u128 = 50000;
-        let curve = StableCurve {
-            amp: 1,
-            ..Default::default()
-        };
-
-        let result = curve
-            .swap(source_amount, swap_source_amount, swap_destination_amount)
-            .unwrap();
-        assert_eq!(result.new_source_amount, 1100);
-        assert_eq!(result.amount_swapped, 2083);
-        assert_eq!(result.new_destination_amount, 47917);
+        const POOL_AMOUNTS: &[u128] = &[
+            100,
+            10_000,
+            1_000_000,
+            100_000_000,
+            10_000_000_000,
+            1_000_000_000_000,
+            100_000_000_000_000,
+            10_000_000_000_000_000,
+            1_000_000_000_000_000_000,
+        ];
+        const SWAP_AMOUNTS: &[u128] = &[
+            100,
+            1_000,
+            10_000,
+            100_000,
+            1_000_000,
+            10_000_000,
+            100_000_000,
+            1_000_000_000,
+            10_000_000_000,
+            100_000_000_000,
+        ];
+        const AMP_FACTORS: &[u64] = &[1, 10, 20, 50, 75, 100, 125, 150];
+
+        for swap_source_amount in POOL_AMOUNTS {
+            for swap_destination_amount in POOL_AMOUNTS {
+                for source_amount in SWAP_AMOUNTS {
+                    for amp in AMP_FACTORS {
+                        let curve = StableCurve {
+                            amp: *amp,
+                            ..Default::default()
+                        };
+
+                        if *source_amount >= *swap_source_amount {
+                            continue;
+                        }
+
+                        println!(
+                            "trying: source_amount={}, swap_source_amount={}, swap_destination_amount={}", 
+                            source_amount, swap_source_amount, swap_destination_amount
+                        );
+
+                        let model: Model = Model::new(
+                            curve.amp.into(),
+                            vec![*swap_source_amount, *swap_destination_amount],
+                            N_COINS,
+                        );
+
+                        let result = curve.swap(
+                            *source_amount,
+                            *swap_source_amount,
+                            *swap_destination_amount,
+                        );
+
+                        let result = result.unwrap();
+                        let sim_result = model.sim_exchange(0, 1, *source_amount);
+
+                        println!(
+                            "result={}, sim_result={}",
+                            result.amount_swapped, sim_result
+                        );
+                        let diff = (sim_result as i128 - result.amount_swapped as i128).abs();
+
+                        assert!(
+                            diff <= 1,
```

Comment:
We should be checking `diff == 0` right? Probably should fix the off-by-one errors?

Reply:
Well, I think it is ok to have errors by 1, after all, this is one lamport error. And we ignore it the rest works great on a wide range of inputs. But this is just my opinion.